### PR TITLE
SwiftSyntax: add a mechanism to define traits of syntax nodes to allow abstract access to popular child kinds. NFC

### PR DESCRIPTION
**Pull request by nkcsgexi**
Swift syntax APIs lack an abstract way of accessing children. The client has to
down-cast a syntax node to the leaf type to access any of its children. However,
some children are common among different syntax kinds, e.g.
DeclAttributeSyntax and DeclMembers. We should allow an abstract way to
access and modify them, so that clients can avoid logic duplication.

This patch adds a mechanism to define new traits and specify satisfied
traits in specific syntax nodes. A trait is a set of common children
and implemented in Swift as a protocol for syntax nodes to conform to.
As a proof-of-concept, we added two traits for now including DeclGroup
and BracedSyntax.

Resolves: #49479 and #49465

### CODE REVIEW

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

File: utils/gyb_syntax_support/Traits.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
+from Child import Child
+
+
+class TRAIT(object):
```
Comment (harlanhaskins) [inline utils/gyb_syntax_support/Traits.py]: Nitpick: `Trait` in title case, please.

**Review by harlanhaskins — APPROVED** [review]
I’ve wanted a nice way to group these and share common nodes. This also makes sure the APIs for things like `leftBrace` are consistently named.

LGTM modulo nitpicks

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

File: utils/gyb_syntax_support/Traits.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+
+class Trait(object):
+    def __init__(self, trait_name, children):
+        self.trait_name = trait_name
+        self.children = children
+
+
+TRAITS = [
+    Trait('DeclGroup',
```
Comment (rintaro) [inline utils/gyb_syntax_support/Traits.py]: I think traits should also be postfixed with "`Syntax`".

Could you make this `DeclGroupSyntax`?
Or make them `DeclGroup` and `Braced`, and postfix in the client?
```swift
public protocol ${trait.trait_name}Syntax {
```

Comment (nkcsgexi) [reply utils/gyb_syntax_support/Traits.py]: Sure, I will rename this to `DeclGroupSyntax`.

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

**Review by rintaro — APPROVED** [review]
Thanks!

**Pull request merged**